### ISSUE TITLE
Migrate to a federated plugin architecture

### ISSUE THREAD

**Issue by mbtodorov** [launchdarkly/flutter-client-sdk#29]
_This relates to but is not a duplicate of #6_

Flutter best practices dictate that plugins should be built using a federated architecture. From the flutter [docs](https://docs.flutter.dev/development/packages-and-plugins/developing-packages#federated-plugins):

> A federated plugin requires the following packages:
> - app-facing package
The package that plugin users depend on to use the plugin. This package specifies the API used by the Flutter app.
> - platform package(s)
One or more packages that contain the platform-specific implementation code. The app-facing package calls into these packages—they aren’t included into an app, unless they contain platform-specific functionality accessible to the end user.
> - platform interface package
The package that glues the app-facing packing to the platform package(s). This package declares an interface that any platform package must implement to support the app-facing package. Having a single package that defines this interface ensures that all platform packages implement the same functionality in a uniform way.

Migrating the launch darkly client SDK to the federated plugin architecture would be extremely beneficial for those of us that would benefit from having web support. For instance, if we only require a handful of the methods available on the `LDClient` class, adding a web platform implementation by leveraging Dart's JS interop with the [dart:js library](https://api.dart.dev/stable/2.19.4/dart-js/dart-js-library.html) and launch darkly's [js sdk](https://github.com/launchdarkly/js-client-sdk), can be a fairly trivial task. This would at the very least unlock the ability to build out an MVP web support for those of us who really need it.

With the plugin's current monolithic architecture, adding support for an additional platform is a very cumbersome task, especially if a third-party would like to implement it.

I would appreciate anyone considering this request 👍.

**Comment by tanderson-ld** [comment]
Hi @mbtodorov, we have had discussions about the platforms supported by the Flutter SDK and also agree we want to improve it for all our Flutter devs.  At the moment we can’t commit to any timeline, but we do hear you!

**Comment by mbtodorov** [comment]
Thanks for the prompt reply @tanderson-ld.

I appreciate that supporting an additional platform is no easy feat. Especially in this case, as I have worked with the LD web SDKs, and am well aware that the APIs and behaviours differ slightly to those of the mobile SDKs. Just to be clear though, the issues is _not_ requesting web support, I didn't mean to open a duplicate of #6. Instead, migrating to a federated plugin architecture just for the existing platforms (android & iOS) would already open the door for the community to make tweaks as needed.

To give you a concrete example, in our company we would largely benefit from an MVP web solution where the only available features are:
- identifying users (anonymous or unique user IDs) 
- evaluating boolean flags (without subscribing to real time changes)

To achieve the above, we just need to provide an implementation for the following methods on the web platform:
- `LDClient.start`
- `LDClient.identify`
- `LDClient.boolVariation`

If this plugin was built using as a federated plugin, we could easily create a [non-endorsed](https://docs.flutter.dev/development/packages-and-plugins/developing-packages#non-endorsed-federated-plugin) implementation for the web platform, where we partially implement the above methods using LD's [js-client-sdk](https://github.com/launchdarkly/js-client-sdk). Since in our use case, we are willing to work with a very relaxed set of features, we can get an implementation out much faster than your teams can add full web support, which again, is no easy feat.

To sum up, if your team has a long-term milestone of adding web support for the flutter SDK, this can be approached iteratively. The first step would be migrate to a federated plugin architecture, which would already unlock the possibility for companies like ours to look and add basic web implementation, while your team is working on the official one.

Hopefully this makes sense, let me know if you would benefit from further elaboration or use case examples. Looking forward to hear your thoughts 👀!

**Comment by kinyoklion** [comment]
Hello,

Version 4.0+ of the SDK no longer has native implementation requirements, and it works on Android, iOS, Windows, macOS, Linux and web. If there becomes a situation where we require platform specific code again, then we will consider this.

Thank you,
Ryan
